ui-manchette-with-spacetimechart/src/hooks/useManchetteWithSpaceTimeChart.ts

```diff
@@ -0,0 +1,194 @@
+    operationalPointsToDisplay: [],
+    panY: false,
+    scales: [],
+  });
+
+  const { xZoom, yZoom, xOffset, yOffset, panY, panning, scrollPosition, isProportional } = state;
+
+  const paths = usePaths(projectPathTrainResult, selectedProjection);
+
+  const checkOverflow = useCallback((isOverflowFromCallback: boolean) => {
+    setState((prev) => ({ ...prev, panY: isOverflowFromCallback }));
+  }, []);
+  useIsOverflow(manchette, checkOverflow);
+
+  // Memoize timeWindow to avoid recalculation on each render
+  const timeWindow = useMemo(
+    () => computeTimeWindow(projectPathTrainResult),
+    [projectPathTrainResult]
+  );
+
+  // Memoize computedOperationalPoints to avoid recalculations unless the dependencies change
+  const computedOperationalPointsToDisplay = useMemo(
+    () => calcOperationalPointsToDisplay(operationalPoints, height, isProportional, yZoom),
+    [operationalPoints, height, isProportional, yZoom]
+  );
+  const computedoperationnalPointsHeight = useMemo(
+    () =>
+      operationalPointsHeight(computedOperationalPointsToDisplay, height, yZoom, isProportional),
+    [computedOperationalPointsToDisplay, height, yZoom, isProportional]
+  );
+
+  const operationalPointsWithPosition = useMemo(
+    () => getOperationalPointsWithPosition(computedOperationalPointsToDisplay),
+    [computedOperationalPointsToDisplay]
+  );
+
+  const zoomYIn = useCallback(() => {
+    if (yZoom < MAX_ZOOM_Y) {
+      setState((prev) => ({ ...prev, yZoom: yZoom + ZOOM_Y_DELTA }));
+    }
+  }, [yZoom]);
+
+  const zoomYOut = useCallback(() => {
+    if (yZoom > MIN_ZOOM_Y) {
+      setState((prev) => ({ ...prev, yZoom: yZoom - ZOOM_Y_DELTA }));
+    }
+  }, [yZoom]);
+
+  const handleScroll = useCallback(() => {
+    if (!isShiftPressed && manchette.current) {
+      const { scrollTop } = manchette.current;
+      if (scrollTop || scrollTop === 0) {
+        setState((prev) => ({ ...prev, scrollPosition: scrollTop, yOffset: scrollTop }));
+      }
+    }
+  }, [isShiftPressed]);
+
+  const handleKeyDown = useCallback((event: KeyboardEvent) => {
+    if (event.key === 'Shift') {
+      setIsShiftPressed(true);
+    }
+  }, []);
+
+  const handleKeyUp = useCallback((event: KeyboardEvent) => {
+    if (event.key === 'Shift') {
+      setIsShiftPressed(false);
+    }
+  }, []);
+
+  useEffect(() => {
+    window.addEventListener('keydown', handleKeyDown);
+    window.addEventListener('keyup', handleKeyUp);
+
+    return () => {
+      window.removeEventListener('keydown', handleKeyDown);
+      window.removeEventListener('keyup', handleKeyUp);
+    };
+  }, [handleKeyDown, handleKeyUp]);
+
+  const toggleMode = useCallback(() => {
+    setState((prev) => ({ ...prev, isProportional: !prev.isProportional }));
+  }, []);
+
+  const computedScales = useMemo(
+    () => getScales(operationalPointsWithPosition, height, isProportional, yZoom),
+    [operationalPointsWithPosition, height, isProportional, yZoom]
+  );
+
+  return {
```

Comment:
A bit of memoization could be nice here, because the object for `manchetteProps` and `spaceTimeChartProps` are always both regenerated, even if only one of them has actual new values. It's possible that this does not occur too often, but still, I would memoize those two objects.

Reply:
Done thanks !